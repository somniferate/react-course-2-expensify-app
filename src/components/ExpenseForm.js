import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calanderFocus: false,
            showError: false
        };
    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/g)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calanderFocus: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ showError: "Please provide both description and amount." }));
        } else {
            this.setState(() => ({ showError: "" }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        };
    };

    render() {
        return (
                <form className="form" onSubmit={this.onSubmit}>
                    {!!this.state.showError && <p className="form__error">{this.state.showError}</p>}
                    <input
                        className="input-text"
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        className="input-text"
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calanderFocus}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        hideKeyboardShortcutsPanel={true}
                    />
                    <textarea
                        className="textarea"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <div>
                        <button className="button">{this.props.expense ? "Save" : "Add"} Expense</button>
                    </div>
                </form>
        );
    };
};

export default ExpenseForm