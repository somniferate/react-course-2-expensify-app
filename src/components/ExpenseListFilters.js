import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters"
import { DateRangePicker } from "react-dates"
import uuid from "uuid"

export class ExpenseListFilters extends React.Component {
    state = {
        focusedInput: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setEndDate(endDate);
        this.props.setStartDate(startDate);
    };

    onFocusChange = (focusedInput) => {
        this.setState({ focusedInput })
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onFilterChange = (e) => {
        e.target.value === "date" ?
            this.props.sortByDate() :
            this.props.sortByAmount()
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="input-text"
                            value={this.props.filters.text}
                            placeholder="Search Expenses"
                            onChange={(e) => this.onTextChange(e)} />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="selector"
                            value={this.props.filters.sortBy}
                            onChange={(e) => this.onFilterChange(e)}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={uuid()}
                            endDate={this.props.filters.endDate}
                            endDateId={uuid()}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.focusedInput}
                            onFocusChange={this.onFocusChange}
                            isOutsideRange={() => false}
                            numberOfMonths={1}
                            showClearDates={true}
                        />
                    </div>
                </div>

            </div>



        )
    };
};

const mapStateToProps = (state) => ({
    filters: state.filter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => dispatch(setTextFilter(value)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)