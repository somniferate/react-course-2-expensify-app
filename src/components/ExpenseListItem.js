import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral"


export const ExpenseListItem = ( { description, amount, createdAt, id } ) => (
    <div>
    <Link to={`/edit/${id}`}><h2>{description}</h2></Link>
        <p>
            {numeral(amount/100).format("$0,0.00")}
            - 
            {moment(createdAt).format("Do MMM, YYYY")}
        </p>
    </div>
);

export default ExpenseListItem