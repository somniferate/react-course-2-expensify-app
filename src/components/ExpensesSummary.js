import React from "react"
import { connect } from "react-redux"
import totalExpenses from "../selectors/expenses-total"
import selectExpense from "../selectors/expenses"
import numeral from "numeral"

export const ExpensesSummary = (props) => {
    const expensesCount = props.expenses.length;
    const expensesTotal = totalExpenses(props.expenses)
    return (
        <div>
            {props.expenses.length !== 0 && <p>{`Viewing ${expensesCount} expense${expensesCount > 1 ? "s" : ""} totalling ${numeral(expensesTotal / 100).format("$0,0.00")}`}</p> }
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filter)
    };
}

export default connect(mapStateToProps)(ExpensesSummary)