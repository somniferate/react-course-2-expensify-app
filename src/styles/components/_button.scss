.button {
    background: lightseagreen;
    border: 0;
    color:white;
    display: inline-block;
    font-size:$ml-size;
    font-weight: 300;
    line-height: 1;
    outline: none;
    padding: $s-size;
    text-decoration: none;
}

.button--link {
    background: none;
}

.button--secondary {
    background: #888;
}