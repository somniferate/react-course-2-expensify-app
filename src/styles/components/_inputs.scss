.input-text {
    border: 1px solid #cacccd;
    height: 50px;
    font-size: $ml-size;
    font-weight: 300;
    padding: $s-size;
}

.selector {
    @extend .input-text;
}

.textarea {
    @extend .input-text;
    height: 10rem;
}