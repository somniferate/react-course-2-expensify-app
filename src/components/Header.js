import React from "react";
import { NavLink } from "react-router-dom"
import ExpensesSummary from "./ExpensesSummary"


const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <ExpensesSummary />
    </header>
);

export default Header;