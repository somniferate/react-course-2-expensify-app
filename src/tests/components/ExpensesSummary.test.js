import React from "react"
import { ExpensesSummary } from "../../components/ExpensesSummary"
import expensesFixtures from "../fixtures/expenses"
import {shallow} from "enzyme"

test("expect nothing to render if zero expenses passed", () => {
    const expenses = []
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test("expect a total to render if one expense passed", () => {
    const expenses = [expensesFixtures[1]]
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test("expect fixtures total to render if all expenses passed", () => {
    const expenses = expensesFixtures
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})