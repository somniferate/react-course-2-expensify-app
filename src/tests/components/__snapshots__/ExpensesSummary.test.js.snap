// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect a total to render if one expense passed 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1,950.00
  </p>
</div>
`;

exports[`expect fixtures total to render if all expenses passed 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $2,351.95
  </p>
</div>
`;

exports[`expect nothing to render if zero expenses passed 1`] = `<div />`;
